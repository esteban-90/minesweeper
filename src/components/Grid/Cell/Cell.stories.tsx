import type { Story, Meta } from '@storybook/react'
import { Cell, type CellProps } from './Cell'

export default {
  title: 'Components/Grid/Cell',
  component: Cell,
  argTypes: {
    coords: {
      defaultValue: [1, 1],
    },
  },
} as Meta

const Template: Story<CellProps> = (args) => <Cell {...args} />

export const ClosedCell = Template.bind({})
ClosedCell.args = { cell: 10 }

export const EmptyCell = Template.bind({})
EmptyCell.args = { cell: 0 }

export const CellWithBomb = Template.bind({})
CellWithBomb.args = { cell: 9 }

export const CellWithFlag = Template.bind({})
CellWithFlag.args = { cell: 11 }

export const CellWithWeakFlag = Template.bind({})
CellWithWeakFlag.args = { cell: 12 }

export const CellWith1 = Template.bind({})
CellWith1.args = { cell: 1 }

export const CellWith3 = Template.bind({})
CellWith3.args = { cell: 3 }

export const CellWith5 = Template.bind({})
CellWith5.args = { cell: 5 }

export const CellWith8 = Template.bind({})
CellWith8.args = { cell: 8 }
