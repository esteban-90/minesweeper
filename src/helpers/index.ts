export type Cell = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12
export type Field = Cell[][]
export type Coords = [number, number]
export type NeighbourItems = Record<string, [number, number]>
export type CellState = Record<string, Cell>

export const getNeighbourItems = ([y, x]: Coords): NeighbourItems => {
  return {
    top: [y - 1, x],
    topRight: [y - 1, x + 1],
    right: [y, x + 1],
    rightBottom: [y + 1, x + 1],
    bottom: [y + 1, x],
    bottomLeft: [y + 1, x - 1],
    left: [y, x - 1],
    leftTop: [y - 1, x - 1],
  }
}

export const checkItemInField = ([y, x]: Coords, { length }: Field): boolean => {
  return y >= 0 && x >= 0 && length - y > 0 && length - x > 0
}

export const incrementNeighbours = (coords: Coords, field: Field): Field => {
  const items = getNeighbourItems(coords)

  for (const item of Object.values(items)) {
    if (checkItemInField(item, field)) {
      const [y, x] = item
      const cell = field[y][x]

      if (cell < 8) field[y][x] = (cell + 1) as Cell
    }
  }

  return field
}

export const cellState: CellState = {
  empty: 0,
  bomb: 9,
  hidden: 10,
  flag: 11,
  weakFlag: 12,
}

export const emptyFieldGenerator = (size: number, state: Cell = cellState.empty): Field => {
  return new Array(size).fill(null).map(() => new Array(size).fill(state))
}

export const errorText = 'Probability must be between 0 and 1'

export const fieldGenerator = (size: number, probability: number): Field => {
  if (probability < 0 || probability > 1) throw new Error(errorText)

  let unprocessedCells = size ** 2
  let restCellsWithBombs = unprocessedCells * probability
  const result = emptyFieldGenerator(size)

  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      if (restCellsWithBombs / unprocessedCells > Math.random()) {
        result[y][x] = cellState.bomb
        incrementNeighbours([y, x], result)
        restCellsWithBombs--
      }

      unprocessedCells--
    }
  }

  return result
}
