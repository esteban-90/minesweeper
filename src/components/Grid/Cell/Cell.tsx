import type { FC, HTMLAttributes, MouseEventHandler } from 'react'
import styled from '@emotion/styled'
import type { Cell as CellType, Coords } from '@/helpers'
import { cellState } from '@/helpers'
import { usePressedButton } from '@/hooks'

export interface CellProps {
  /**
   * Cell status based on cell type
   */
  cell: CellType
  /**
   * Cell coordinates
   */
  coords: Coords
  /**
   * Click handler
   */
  leftClick: (coords: Coords) => void
  /**
   * Context menu handler
   */
  rightClick: (coords: Coords) => void
  /**
   * For testing
   */
  'data-testid'?: string
}

export const checkIsActiveCell = (cell: CellType): boolean => {
  return [cellState.hidden, cellState.flag, cellState.weakFlag].includes(cell)
}

export const Cell: FC<CellProps> = ({ cell, coords, leftClick, rightClick, 'data-testid': id }): JSX.Element => {
  const [pressed, onMouseDown, onMouseUp] = usePressedButton()

  const onClick: MouseEventHandler<HTMLDivElement> = () => {
    leftClick(coords)
  }

  const onContextMenu: MouseEventHandler<HTMLDivElement> = (event) => {
    event.preventDefault()
    if (checkIsActiveCell(cell)) rightClick(coords)
  }

  type StyledCellProps = HTMLAttributes<HTMLDivElement> & { 'data-testid': typeof id }

  const activeCellProps: StyledCellProps & { pressed: typeof pressed } = {
    onClick,
    onContextMenu,
    onMouseDown,
    onMouseUp,
    onMouseLeave: onMouseUp,
    'data-testid': id,
    pressed,
  }

  const nonActiveCellProps: StyledCellProps = {
    onContextMenu,
    'data-testid': id,
  }

  const { empty, bomb, flag, weakFlag, hidden } = cellState

  switch (cell) {
    case empty: {
      return <RevealedCell {...nonActiveCellProps} />
    }

    case bomb: {
      return (
        <CellWithBomb {...nonActiveCellProps}>
          <Bomb />
        </CellWithBomb>
      )
    }

    case hidden: {
      return <ClosedCell {...activeCellProps} />
    }

    case flag: {
      return (
        <ClosedCell {...activeCellProps}>
          <Flag />
        </ClosedCell>
      )
    }

    case weakFlag: {
      return (
        <ClosedCell {...activeCellProps}>
          <WeakFlag />
        </ClosedCell>
      )
    }

    default: {
      return <RevealedCell {...nonActiveCellProps}>{cell}</RevealedCell>
    }
  }
}

export const ClosedCell = styled.div<{ pressed?: boolean }>`
  display: flex;
  align-items: center;
  justify-content: center;
  user-select: none;
  cursor: pointer;
  width: 1.8vw;
  height: 1.8vw;
  background-color: #d1d1d1;
  border: 0.6vh solid transparent;
  border-color: ${({ pressed = false }) => (pressed ? 'tranparent' : 'white #9e9e9e #9e9e9e white')};

  &:hover {
    filter: brightness(1.1);
  }
`

const RevealedCell = styled(ClosedCell)`
  border-color: #ddd;
  cursor: default;
  color: ${({ children }) => colors[children as CellType] ?? transparent};

  &:hover {
    filter: brightness(1);
  }
`

const Bomb = styled.div`
  border-radius: 50%;
  width: 1vw;
  height: 1vw;
  background-color: #333;
`

const CellWithBomb = styled(RevealedCell)`
  background-color: #ec433c;
`

const Flag = styled.div`
  width: 0;
  height: 0;
  border-top: 0.5vw solid transparent;
  border-bottom: 0.5vw solid transparent;
  border-left: 0.5vw solid #ec433c;
`

const WeakFlag = styled(Flag)`
  border-left: 0.5vw solid #f19996;
`

const transparent = 'rgba(0,0,0,0)'

const colors: { [key in CellType]: string } = {
  0: '#000',
  1: '#2a48ec',
  2: '#2bb13d',
  3: '#ec6561',
  4: '#233db7',
  5: '#a6070f',
  6: '#e400af',
  7: '#906a02',
  8: '#fa0707',
  9: transparent,
  10: transparent,
  11: transparent,
  12: transparent,
}
