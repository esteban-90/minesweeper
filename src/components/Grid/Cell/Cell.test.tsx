import { render, screen, fireEvent, createEvent } from '@testing-library/react'
import { cellState, type Coords } from '@/helpers'
import { Cell, type CellProps, checkIsActiveCell } from './Cell'

describe('Cell test cases:', () => {
  const coords: Coords = [1, 1]
  const { empty, weakFlag } = cellState
  const mockFn = vitest.fn()

  for (let cell = empty; cell <= weakFlag; cell++) {
    const id = `${cell}_${coords}`
    const props: CellProps = { coords, leftClick: mockFn, rightClick: mockFn, cell, 'data-testid': id }

    beforeEach(() => {
      render(<Cell {...props} />)
    })

    it(`should check prevent default onContextMenu for type of cell "${cell}"`, () => {
      const cellElement = screen.getByTestId(id)
      const contextMenuEvent = createEvent.contextMenu(cellElement)

      fireEvent(cellElement, contextMenuEvent)
      expect(contextMenuEvent.defaultPrevented).toBe(true)
    })

    it(`onClick and onContextMenu handlers should be called for active cell "${cell}"`, () => {
      const isActive = checkIsActiveCell(cell)
      const cellElement = screen.getByTestId(id)

      fireEvent.click(cellElement)
      fireEvent.contextMenu(cellElement)

      if (isActive) {
        expect(props.leftClick).toBeCalled()
        expect(props.rightClick).toBeCalled()
      } else {
        expect(props.leftClick).not.toBeCalled()
        expect(props.rightClick).not.toBeCalled()
      }
    })
  }
})
