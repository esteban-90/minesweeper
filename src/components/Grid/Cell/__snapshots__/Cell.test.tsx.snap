// Vitest Snapshot v1

exports[`Cell test cases: > renders: > cell with status "0" should render correctly 1`] = `
<DocumentFragment>
  <div
    class="css-1ohgs2d"
    data-testid="0_1,1"
  />
</DocumentFragment>
`;

exports[`Cell test cases: > renders: > cell with status "1" should render correctly 1`] = `
<DocumentFragment>
  <div
    class="css-kj4uv0"
    data-testid="1_1,1"
  >
    1
  </div>
</DocumentFragment>
`;

exports[`Cell test cases: > renders: > cell with status "2" should render correctly 1`] = `
<DocumentFragment>
  <div
    class="css-6f1wo7"
    data-testid="2_1,1"
  >
    2
  </div>
</DocumentFragment>
`;

exports[`Cell test cases: > renders: > cell with status "3" should render correctly 1`] = `
<DocumentFragment>
  <div
    class="css-1ild62n"
    data-testid="3_1,1"
  >
    3
  </div>
</DocumentFragment>
`;

exports[`Cell test cases: > renders: > cell with status "4" should render correctly 1`] = `
<DocumentFragment>
  <div
    class="css-v3wr85"
    data-testid="4_1,1"
  >
    4
  </div>
</DocumentFragment>
`;

exports[`Cell test cases: > renders: > cell with status "5" should render correctly 1`] = `
<DocumentFragment>
  <div
    class="css-nheh0a"
    data-testid="5_1,1"
  >
    5
  </div>
</DocumentFragment>
`;

exports[`Cell test cases: > renders: > cell with status "6" should render correctly 1`] = `
<DocumentFragment>
  <div
    class="css-1ttleus"
    data-testid="6_1,1"
  >
    6
  </div>
</DocumentFragment>
`;

exports[`Cell test cases: > renders: > cell with status "7" should render correctly 1`] = `
<DocumentFragment>
  <div
    class="css-1p2mldb"
    data-testid="7_1,1"
  >
    7
  </div>
</DocumentFragment>
`;

exports[`Cell test cases: > renders: > cell with status "8" should render correctly 1`] = `
<DocumentFragment>
  <div
    class="css-a2fdva"
    data-testid="8_1,1"
  >
    8
  </div>
</DocumentFragment>
`;

exports[`Cell test cases: > renders: > cell with status "9" should render correctly 1`] = `
<DocumentFragment>
  <div
    class="css-77ydc3"
    data-testid="9_1,1"
  >
    <div
      class="css-3tubcw"
    />
  </div>
</DocumentFragment>
`;

exports[`Cell test cases: > renders: > cell with status "10" should render correctly 1`] = `
<DocumentFragment>
  <div
    class="css-5803oy"
    data-testid="10_1,1"
  />
</DocumentFragment>
`;

exports[`Cell test cases: > renders: > cell with status "11" should render correctly 1`] = `
<DocumentFragment>
  <div
    class="css-5803oy"
    data-testid="11_1,1"
  >
    <div
      class="css-oefih3"
    />
  </div>
</DocumentFragment>
`;

exports[`Cell test cases: > renders: > cell with status "12" should render correctly 1`] = `
<DocumentFragment>
  <div
    class="css-5803oy"
    data-testid="12_1,1"
  >
    <div
      class="css-jm2piq"
    />
  </div>
</DocumentFragment>
`;

exports[`Cell test cases: > renders: > closed cell should render correctly 1`] = `
<DocumentFragment>
  <div
    class="css-12l1qmc"
  />
</DocumentFragment>
`;

exports[`Cell test cases: > renders: > closed cell should render correctly 2`] = `
<DocumentFragment>
  <div
    class="css-12l1qmc"
  />
</DocumentFragment>
`;

exports[`Cell test cases: > renders: > closed cell should render correctly 3`] = `
<DocumentFragment>
  <div
    class="css-12l1qmc"
  />
</DocumentFragment>
`;

exports[`Cell test cases: > renders: > closed cell should render correctly 4`] = `
<DocumentFragment>
  <div
    class="css-12l1qmc"
  />
</DocumentFragment>
`;

exports[`Cell test cases: > renders: > closed cell should render correctly 5`] = `
<DocumentFragment>
  <div
    class="css-12l1qmc"
  />
</DocumentFragment>
`;

exports[`Cell test cases: > renders: > closed cell should render correctly 6`] = `
<DocumentFragment>
  <div
    class="css-12l1qmc"
  />
</DocumentFragment>
`;

exports[`Cell test cases: > renders: > closed cell should render correctly 7`] = `
<DocumentFragment>
  <div
    class="css-12l1qmc"
  />
</DocumentFragment>
`;

exports[`Cell test cases: > renders: > closed cell should render correctly 8`] = `
<DocumentFragment>
  <div
    class="css-12l1qmc"
  />
</DocumentFragment>
`;

exports[`Cell test cases: > renders: > closed cell should render correctly 9`] = `
<DocumentFragment>
  <div
    class="css-12l1qmc"
  />
</DocumentFragment>
`;

exports[`Cell test cases: > renders: > closed cell should render correctly 10`] = `
<DocumentFragment>
  <div
    class="css-12l1qmc"
  />
</DocumentFragment>
`;

exports[`Cell test cases: > renders: > closed cell should render correctly 11`] = `
<DocumentFragment>
  <div
    class="css-12l1qmc"
  />
</DocumentFragment>
`;

exports[`Cell test cases: > renders: > closed cell should render correctly 12`] = `
<DocumentFragment>
  <div
    class="css-12l1qmc"
  />
</DocumentFragment>
`;

exports[`Cell test cases: > renders: > closed cell should render correctly 13`] = `
<DocumentFragment>
  <div
    class="css-12l1qmc"
  />
</DocumentFragment>
`;
